### PR TITLE
add the ability to remove nodes older than a certain age even when the node group has reached the min

### LINKED ISSUE #239: [Feature Request] Add ability to remove nodes older than a certain amount of time
**Is your feature request related to a problem? Please describe.**
We have a number of node groups that receive a small amount of load and not enough to warrant more than 1 instance. As a result, the node in these node groups stay around in the cluster for a long time and do not receive AMI/launch template/security updates.

**Describe the solution you'd like**
Have Escalator remove old nodes automatically when the age of them exceeds a pre-configured value. 

**Describe alternatives you've considered**
We normally resolve this by bumping the ASG desired size to a larger number to bring up new nodes and for Escalator to terminate the nodes as they are the oldest.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by awprice**
- adds max_node_age field to node group configuration
- max_node_age by default is disabled, and needs to be a positive, greater than zero duration to enable
- when the node group has reached the minimum and there are nodes older than the max_node_age, the scale delta will be set to +1
- this will then force escalator to replace the oldest node

fixes #239

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -431,6 +437,33 @@ func (c *Controller) isScaleOnStarve(
...
+// when the node group is at the minimum. This is to ensure the nodes are receiving the most up-to-date configuration
+// from the cloud provider.
+func (c *Controller) scaleOnMaxNodeAge(nodeGroup *NodeGroupState, untaintedNodes []*v1.Node) bool {
+	// Only enable this functionality if the node group has the feature enabled
+	if nodeGroup.Opts.MaxNodeAgeDuration() <= 0 {
+		return false
+	}
+
+	// We don't want to attempt to rotate nodes that have reached the max age if we haven't reached the minimum node
+	// count, as the scaling down of the node group will remove the oldest first anyway.
```
Comment (mwhittington21) [inline pkg/controller/controller.go:451]: There's still an edge case here which is a node group that somehow has very consistent load such that it always maintains X number of nodes (e.g. always 2) and thus never gets cycled. I think this is very unlikely with the kinds of workloads that run on Escalator, but worth calling out in case others think it's worth considering here.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -431,6 +437,33 @@ func (c *Controller) isScaleOnStarve(
...
+	if len(untaintedNodes) != nodeGroup.Opts.MinNodes || len(untaintedNodes) == 0 {
```
Comment (mwhittington21) [inline pkg/controller/controller.go:452]: Should this be <= to catch weird edge cases?

File: pkg/controller/controller_scale_node_group_test.go [hunk 3]
```diff
@@ -1060,3 +1060,211 @@ func TestScaleNodeGroup_MultipleRuns(t *testing.T) {
...
+					Name:                    "default",
+					CloudProviderGroupName:  "default",
+					MinNodes:                3,
+					MaxNodes:                10,
+					ScaleUpThresholdPercent: 70,
+					MaxNodeAge:              "12h",
+				},
+				listerOptions: ListerOptions{},
+			},
+			1,
```
Comment (mwhittington21) [inline pkg/controller/controller_scale_node_group_test.go:1128]: So in this case, the behaviour is that we scale up one node, it would be replaced and then we would do another scale up to replace the other older node once we made it back to our minimum node count again. Is that correct?

File: pkg/controller/controller_scale_node_group_test.go [hunk 4]
```diff
@@ -1060,3 +1060,211 @@ func TestScaleNodeGroup_MultipleRuns(t *testing.T) {
...
+					ScaleUpThresholdPercent: 70,
+					MaxNodeAge:              "12h",
+				},
+				listerOptions: ListerOptions{},
+			},
+			0,
+			nil,
+		},
+		{
+			"max_node_age enabled, some nodes are tainted",
```
Comment (mwhittington21) [inline pkg/controller/controller_scale_node_group_test.go:1194]: Should there be a test case where we're at the MinNodes but some are tainted? I don't think that should occur in regular operation though, just throwing this out there in case it makes sense.

File: pkg/controller/node_group_test.go [hunk 5]
```diff
@@ -471,6 +476,31 @@ func TestValidateNodeGroup(t *testing.T) {
 					HardDeleteGracePeriod:              "1h10m",
 					ScaleUpCoolDownPeriod:              "55m",
 					TaintEffect:                        "",
+					MaxNodeAge:                         "",
```
Comment (mwhittington21) [inline pkg/controller/node_group_test.go:479]: Piggy backing this particular edge case onto the case where there's an empty TaintEffect means that us also checking that we're parsing an empty MaxNodeAge is kinda lost. Either make a new table test case here or add that we're checking this field in the name of the table test case.

File: pkg/controller/controller_scale_node_group_test.go [hunk 6]
```diff
@@ -1060,3 +1060,211 @@ func TestScaleNodeGroup_MultipleRuns(t *testing.T) {
...
+		nodeGroupOptions NodeGroupOptions
+		listerOptions    ListerOptions
+	}
+
+	tests := []struct {
+		name              string
+		args              args
+		expectedNodeDelta int
+		err               error
+	}{
```
Comment (Jacobious52) [inline pkg/controller/controller_scale_node_group_test.go:1086]: could we add a test for num nodes == maxNodes, and all expired, with scale up enabled

File: pkg/controller/controller.go (hunk 1 above)
Comment (awprice) [reply pkg/controller/controller.go:451]: I had thought of this edge case whilst writing this feature - I guess it's a been an edge case in Escalator since it's existence, and won't be fixed or made worse by this feature being added.

Ideally if there is a consistent base load, the mininum nodes in the ASG should really be configured to a value that meets the base load, then this feature can be used to ensure the nodes are cycled out.

I can check if this edge case is documented in Escalator's documentation

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -373,6 +373,12 @@ func (c *Controller) scaleNodeGroup(nodegroup string, nodeGroup *NodeGroupState)
 		nodesDelta = int(math.Max(float64(nodesDelta), 1))
 	}
 
+	if c.scaleOnMaxNodeAge(nodeGroup, untaintedNodes) {
+		log.WithField("nodegroup", nodegroup).
+			Info("Setting scale to minimum of 1 to rotate out a node older than the max node age")
+		nodesDelta = int(math.Max(float64(nodesDelta), 1))
```
Comment (Jacobious52) [inline pkg/controller/controller.go:379]: should this also protect against nodesDelta exceeding ASGMax? Or is it protected against in normal scale functions

File: pkg/controller/controller_scale_node_group_test.go (hunk 6 above)
Comment (Jacobious52) [reply pkg/controller/controller_scale_node_group_test.go:1086]: yeah just the very unlikely case of nodes at max for a long time, and this min of +1 node delta to cycle them trying to push above maxASG.

File: pkg/controller/controller.go (hunk 2 above)
Comment (awprice) [reply pkg/controller/controller.go:452]: No I think we want to keep `len(untaintedNodes) != nodeGroup.Opts.MinNodes`, as we're only wanting this feature to come into play when we're at the minimum node level.

File: pkg/controller/controller_scale_node_group_test.go (hunk 3 above)
Comment (awprice) [reply pkg/controller/controller_scale_node_group_test.go:1128]: Yep - that's the desired behaviour

File: pkg/controller/controller_scale_node_group_test.go (hunk 6 above)
Comment (awprice) [reply pkg/controller/controller_scale_node_group_test.go:1086]: It's actually perfectly fine for node delta to be +1 when at the max ASG size - we have logic later on when trying to increase the ASG in the cloud provider above the max that will reject the scale up.

I don't think this test case is really relevant for this specific feature and the case of node delta exceeding the ASG max should be tested elsewhere.

Ideally we're really testing if nodeDelta is either +1 or 0 given the criteria for the feature and assuming whatever Escalator does with that nodeDelta is appropriately tested later on

File: pkg/controller/controller.go (hunk 7 above)
Comment (awprice) [reply pkg/controller/controller.go:379]: This is somewhat related to the other conversation, but I don't think this is relevant for the tests of this feature - this should already be covered by the general scaling tests that check that if nodeDelta would breach the ASG max, that Escalator prevents it.

File: pkg/controller/controller.go (hunk 2 above)
Comment (mwhittington21) [reply pkg/controller/controller.go:452]: What I meant was using `<=` is more defensive in case we (or the cloud provider) have a bug elsewhere in the code that would allow the number of nodes present to be less than the number of MinNodes. In that case, we would actually start checking their age and potentially mess with the scale delta here, which complicates getting back to the MinNodes regardless.

All of this said, it's an unlikely edge case and I'm fine with this.

**Review by mwhittington21 — DISMISSED** [review]
lgtm

File: pkg/controller/controller.go (hunk 1 above)
Comment (awprice) [reply pkg/controller/controller.go:451]: > I can check if this edge case is documented in Escalator's documentation

This isn't documented anywhere - I've raised #241

File: pkg/controller/controller.go (hunk 2 above)
Comment (awprice) [reply pkg/controller/controller.go:452]: > What I meant was using <= is more defensive in case we (or the cloud provider) have a bug elsewhere in the code that would allow the number of nodes present to be less than the number of MinNodes. 

Exceeding the max and being under the min is already caught by Escalator here - https://github.com/atlassian/escalator/blob/97246b1ff512ecdb000a856f10b673bd1dfcbb8b/pkg/controller/controller.go#L237-L254

**Review by Jacobious52 — APPROVED** [review]

File: pkg/controller/controller_scale_node_group_test.go (hunk 4 above)
Comment (awprice) [reply pkg/controller/controller_scale_node_group_test.go:1194]: Had a look into adding this test case - it doesn't make sense to add to this feature's tests, as Escalator should get its self out of this scenario prior to the maxNodeAge code running (it will untaint the tainted nodes in an attempt to get the untainted nodes equal to the min nodes).

Again, this should be tested elsewhere already.

File: pkg/controller/node_group_test.go (hunk 5 above)
Comment (awprice) [reply pkg/controller/node_group_test.go:501]: Have added a test case for this

**Review by Jacobious52 — APPROVED** [review]

**Review by mwhittington21 — DISMISSED** [review]

**Review by mwhittington21 — APPROVED** [review]

**Review by mwhittington21 — APPROVED** [review]

**Review by Jacobious52 — APPROVED** [review]
